docs/design/features/bridging-constraints.md

```diff
@@ -0,0 +1,110 @@
+        DoConstrained<T>();
+    }
+}
+```
+
+The purpose of this document is to sketch out a possible implementation on the runtime side that would allow above code to work.
+
+This feature doesn't assume any kind of relaxation of constraint checks outside method bodies. E.g. allowing to derive from `class Base<T> where T : IFoo { }` with `class Derived<T> : Base<T> { }` to load for `Derived<IFoo>` is not in scope.
+
+## Runtime impact
+
+There are two ways how constraints are validated within the runtime right now:
+
+* Validating they match on uninstantiated types: done by comparing constraints on generic type parameters - if type parameter T is used to substitute U, the constraints need to be satisfied.
+* Validating they match on concrete instantiation: for example `class Foo<T> where T: struct { }`: `Foo<int>` satisfies constraints, `Foo<object>` doesn't.
+
+Validating on uninstantiated forms is done as part of the old verification logic. The runtime doesn't do this consistently - for example, it is done in calls in IL, but not for ldtoken in IL.
+
+Validating instantiated forms is done always so that a `MethodTable` for an invalid type cannot even be created.
+
+From the runtime perspective, it would be desirable to maintain the invariant that `MethodTable` for a generic type instantiation that doesn't meet its own constraints can never be loaded - an attempt to do so would continue throwing an exception. This has implications on how method bodies that may be violating constraints in certain branches can be compiled. Consider:
+
+```csharp
+//
+// This is pseudo-C# just so that I don't have to write IL.
+// Not expected to compile as C# in this shape, but imagine the straightforward lowering to IL.
+//
+interface IInterface { }
+
+struct Constrained<T> where T : IInterface { }
+
+static void Do<T>()
+{
+    if (typeof(T).IsAssignableFrom(typeof(IInterface))
+    {
+        Constrained<T> someLocal = default;
+        try
+        {
+            someLocal.ToString();
+        }
+        catch (GenericException<Constrained<T>>) { }
+    }
+}
+```
+
+The requirement of never having an invalid type instantiation within the runtime type system means that if we're compiling `Do<int>` (where we cannot allow `Constrained<int>` to exist), the code generator cannot ask questions about all locals (IL locals are scoped to the method, not to a basic block), exception handling regions, or any IL within the `IsAssignableFrom` check. While it may be reasonable to expect the `IsAssignableFrom` check gets optimized out, it would likely not happen for unoptimized code. Locals and EH regions pose additional programs.
+
+It would greatly simplify the codegen impact of this if we were to limit the amount of places within a method body that are allowed to break the constraints. The least impactful and potentially sufficiently powerful change would be allowing to break constraints on `call` instruction.
+
+The proposal is to:
```

Comment:
Was it considered to reuse some of the logic that exists for `static virtuals in interfaces`?

That is, `static virtuals in interfaces` works today via `constrained <type-token> call <method-token>` and the runtime has special logic that allows this to resolve the method and actually function.

Could a similar thing be done here such that the compiler does a relevant check and then simply emits a `constrained call` allowing the runtime to do any additional work required to validate the `type-token` is actually compatible with the required constraints for `method-token`?

Reply:
What would be the `method-token` in this case?

For constrained calls on instance methods, the `type-token` is the type of the `this` used in the interface call. For static methods, it's the type on which to search for the static implementation. I don't see a straightforward thing we could put there for constraint bridging.

We could potentially put there something non-sensical (like `constrained void`) that would serve as a marker, as an alternative to what Jan proposed with `call RuntimeHelpers.CheckConstraintsForMethodCallThatFollows` above, but I don't see anything meaningful we could put there.